### PR TITLE
Check output shape

### PR DESCRIPTION
**Pull request by BruceDai**

### CODE REVIEW

**Comment by BruceDai** [comment]
@huningxin @mingmingtasd @miaobin PTAL, thanks.

File: src/nn/ops/batch_norm.ts [hunk 1]
```diff
@@ -15,6 +15,7 @@ export class BatchNormalization extends SingleOutputOperation implements
   private axis_?: number;
   private epsilon_?: number;
   private activation_?: MLOperator;
+  private needCheckOutputShape_ = true;
```
Comment (mingmingtasd) [inline src/nn/ops/batch_norm.ts:18]: Many ops defined this variable, can it be defined in parent class such as SingleOutputOperation?

Comment (mingmingtasd) [inline src/nn/ops/batch_norm.ts:18]: Many ops defined this variable, can it be defined in parent class such as SingleOutputOperation?

File: src/nn/ops/split.ts [hunk 2]
```diff
@@ -38,6 +39,27 @@ export class Split extends Operation {
...
+      let sliceSizes = [];
+      if (typeof this.splits_ === 'number') {
+        sliceSizes =
+            new Array(this.splits_).fill(input.shape[axis] / this.splits_);
+      } else {
+        sliceSizes = this.splits_.slice();
+      }
+      const outputsShape = [];
+      for (const size of sliceSizes) {
+        const outputShape = input.shape.slice();
```
Comment (mingmingtasd) [inline src/nn/ops/split.ts:54]: Move this line out of the loop.

Comment (mingmingtasd) [inline src/nn/ops/split.ts:54]: Move this line out of the loop.

**Review by mingmingtasd — COMMENTED** [review]
Great work! Thanks ! @BruceDai 
I reviewed for most of ops except the Conv2dTranspose and Gru. Please help review again and especially focus on these two. @miaobin

File: src/nn/ops/batch_norm.ts (hunk 1 above)
Comment (BruceDai) [reply src/nn/ops/batch_norm.ts:18]: Actually, it's good to add this member to its base class. Here, it's on my purpose to remind we need use this `needCheckOutputShape_ ` member to check output shape. @mingmingtasd WDYT?

File: src/nn/ops/binary.ts [hunk 3]
```diff
@@ -23,7 +24,31 @@ export abstract class Binary extends SingleOutputOperation {
   run(inputTensors: Map<MLOperand, tf.Tensor>): tf.Tensor {
     const a: tf.Tensor = inputTensors.get(this.a_);
     const b: tf.Tensor = inputTensors.get(this.b_);
-    return this.runOp(a, b);
+    const output = this.runOp(a, b);
```
[the remark this answers is not in the document]
Comment (BruceDai) [reply src/nn/ops/binary.ts:27]: @mingmingtasd This run function would be invoked twice, the first time is on compiling model phase, the second is on computing phase, I use `needCheckOutputShape_ ` member/flag to just check output shape on compiling model phase, so it doesn't matter with computing phase.

File: src/nn/ops/split.ts (hunk 2 above)
Comment (BruceDai) [reply src/nn/ops/split.ts:54]: @mingmingtasd The `outputShape` would be updated in `for` loop for each sliceSie, so it doesn't need to move out, thanks.

Comment (mingmingtasd) [reply src/nn/ops/split.ts:54]: > 

Need outputShape = input.shape.slice() be updated?

File: src/nn/ops/batch_norm.ts (hunk 1 above)
Comment (mingmingtasd) [reply src/nn/ops/batch_norm.ts:18]: I think code like line103: this.needCheckOutputShape_ = false already remind ,WDYT?

File: src/nn/ops/binary.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mingmingtasd) [reply src/nn/ops/binary.ts:27]: Thanks for your explanation, got it. @BruceDai

**Comment by miaobin** [comment]
> Great work! Thanks ! @BruceDai I reviewed for most of ops except the Conv2dTranspose and Gru. Please help review again and especially focus on these two. @miaobin

Great work! I've meet some issues with formula comprehension and have discussed with @BruceDai LGTM, Thanks!

**Comment by BruceDai** [comment]
Thanks @mingmingtasd and @miaobin for review this patch, let's merge it.

**Pull request merged**